ctapipe/tools/stage1.py

```diff
@@ -0,0 +1,767 @@
+                    tel_index,
+                    params.hillas,
+                    params.timing,
+                    params.leakage,
+                    params.concentration,
+                    params.morphology,
+                    params.intensity,
+                ]
+
+                # currently the HDF5TableWriter has problems if the first event
+                # has NaN as values, since it can't infer the data types.
+                # that implies we need to specify them in the Fields, rather than
+                # infer from first event, perhaps.  For now we skip them.
+                parameters_were_computed = (
+                    False if params.hillas.intensity is np.nan else True
+                )
+
+                if parameters_were_computed:
+                    writer.write(
+                        table_name=f"dl1/event/telescope/parameters/{table_name}",
+                        containers=containers_to_write,
+                    )
+                extra.image_mask = image_mask
+
+            if self.write_images:
+                # note that we always write the image, even if the image quality
+                # criteria are not met (those are only to determine if the parameters
+                # can be computed).
+                writer.write(
+                    table_name=f"dl1/event/telescope/images/{table_name}",
+                    containers=[tel_index, data, extra],
+                )
+
+    def _generate_table_indices(self, h5file, start_node):
+
+        for node in h5file.iter_nodes(start_node):
+            if not isinstance(node, tables.group.Group):
+                self.log.debug(f"gen indices for: {node}")
+                if "event_id" in node.colnames:
+                    node.cols.event_id.create_index()
+                    self.log.debug("generated event_id index")
+                if "tel_id" in node.colnames:
+                    node.cols.tel_id.create_index()
+                    self.log.debug("generated tel_id index")
+                if "obs_id" in node.colnames:
+                    self.log.debug("generated obs_id index")
+                    node.cols.obs_id.create_index(kind="ultralight")
+            else:
+                # recurse
+                self._generate_table_indices(h5file, node)
+
+    def _generate_indices(self, writer):
+
+        if self.write_images:
+            self._generate_table_indices(writer._h5file, "/dl1/event/telescope/images")
+        self._generate_table_indices(writer._h5file, "/dl1/event/subarray")
+
+    def start(self):
+
+        with HDF5TableWriter(
+            self.output_filename, mode="a", add_prefix=True, filters=self._hdf5_filters
+        ) as writer:
+
+            if self.write_parameters is False:
+                # don't need to write out the image mask if no parameters are computed,
+                # since we don't do image cleaning in that case.
+                writer.exclude("/dl1/event/telescope/images", "image_mask")
```

Comment:
Writing DL1a with cleaning applied might be interesting for the deep learning people

Reply:
quite the contrary, the image without cleaning should always be written in DL1a

Reply:
Yes of course, I meant an additional cleaning masked, not the image with pixels set to 0.

Reply:
You can just multiply the mask by the image, so I think it's ok for deep-learning (unless they cannot apply some math to inputs first?).

Reply:
I meant that it is probably nice for them to include the cleaning mask even if no dl1b is written